### ISSUE TITLE
Follow ups to multi-variant rendering

### ISSUE THREAD

**Issue by cmdcolin** [GMOD/jbrowse-components#4712]
- handle polyploid (ALT's more than 0/0,1/0,1/1) better
- allow deleting rows
- allow renaming rows
- add the 'bulk editor' to multi-wiggle
- add clustering method to multi-wiggle
- handle large cnv type data or overlapping features
- add legend
- add mouseover
- add ability to configure metadata better for multivariant. currently requires using the bulk editor

**Comment by cmdcolin** [comment]
example of showing larger scale  SVs in a multi-variant type way, this example with inversion polymorphism (from https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02919-8/figures/1)

![image](https://github.com/user-attachments/assets/6770cb2a-59d2-4162-a1c3-ce1304ba7f5c)

**Comment by cmdcolin** [comment]
could look into more hprc data from 
More here https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC3/release/

**Comment by taprs** [comment]
Any workarounds to display polyploid variant calls so far? I picture it to myself as a heatmap with proportions of ALT calls in each cell (e.g. 0/0/0/1 is at 25% on the color scale). Obviously not applicable to multiallelic sites but they are a headache anyway.

**Comment by cmdcolin** [comment]
>I picture it to myself as a heatmap with proportions of ALT calls in each cell (e.g. 0/0/0/1 is at 25% on the color scale). Obviously not applicable to multiallelic sites but they are a headache anyway.

this is a great idea. I'll see what i could do to try something like that. if you're aware of any good polyploid data to look at, let me know, i think i have some i stumbled on or i can make some synthetic data. 

also, indeed, the code has thus far not done any handling of multi-allelic sites either, sticking to just which it has 4 colors 0/0 (ref, grey) 0/1 (het, teal) and 1/1 (hom, blu) and "other" (purple)

**Comment by taprs** [comment]
Thanks for the quick reply! We have diploid/tetraploid mix data at hand, it's not published yet but we can share it if needed. (Publishing it is exactly why I started looking into the capabilities of Jbrowse2 😉 )

**Comment by cmdcolin** [comment]
that's awesome. I believe I saw a talk from your group at PAG just last week! feel free to email if interested, especially if it has mixed ploidy would be curious, but might be able to find synthetic data too

**Comment by cmdcolin** [comment]
@taprs I did a little work on trying to more properly handle both phased variants and polyploid in this PR

if you are using the web version, you can get the branch build with e.g. `jbrowse create --branch polyploid newinstance`

https://github.com/GMOD/jbrowse-components/pull/4795

if you have phased variants, it has some particular interestingness perhaps because it can split out each haplotypes into a new row


example with a trio vcf

![Screenshot From 2025-01-25 15-03-21](https://github.com/user-attachments/assets/a9308a40-ab74-48c0-9ab6-035b50a1ae0b)

**Comment by taprs** [comment]
Thanks for the news! I set out to test it and caught an error when trying to display a random VCF region in matrix mode (default variant mode is fine, and "regular" multivariant mode does not render tetraploid samples). Here is the stack trace:

```
TypeError: Cannot read properties of undefined (reading 'includes')
TypeError: Cannot read properties of undefined (reading 'includes')
../../../plugins/variants/src/MultiLinearVariantMatrixRenderer/makeImageData.ts:96:33 (at o ()
../../../plugins/variants/src/MultiLinearVariantMatrixRenderer/LinearVariantMatrixRenderer.ts:23:16 (at)
../../../packages/core/util/offscreenCanvasUtils.tsx:72:26 (at l ()
../../../plugins/variants/src/MultiLinearVariantMatrixRenderer/LinearVariantMatrixRenderer.ts:19:7 (at z.render ()
../../../packages/core/util/index.ts:1011:9 (at async he ()
../../../packages/core/pluggableElementTypes/renderers/ServerSideRendererType.tsx:190:11 (at async z.renderInWorker ()
../../../packages/core/rpc/methods/CoreRender.ts:64:37 (at async A.execute ()
JBrowse 2.18.0
```

![Image](https://github.com/user-attachments/assets/4331435e-51c4-4be4-b1ba-ba55f9a1eae0)

I set up the server like this:

```shell
jbrowse create --branch polyploid heatmap_test
jbrowse add-assembly ./NT1_220222.fasta --load symlink -n NT1 --out heatmap_test
jbrowse add-track ./Eur_lyrata_sc1.vcf.gz --load symlink --out heatmap_test
cd heatmap_test
npx serve -S .
```

I can send you the genome + the VCF file link via email this week. I get no error when I display them using the main branch, but understand that this VCF might be special in a few ways:

 - there are both diploids and tetraploids in the same file
 - all genomic positions are there, including sites with no data or with uniform allelic state
 - some regions might have very little to no data

Cheers,
Nikita

**Comment by cmdcolin** [comment]
interesting, i would be curious to see the VCF. that is basically completely failing to parse the genotypes out of the vcf if so. 

if you can show just like a single line, i might be able to extrapolate, but potentially might need the whole vcf also

**Comment by taprs** [comment]
Sure, here are some 20 VCF lines stripped from the region shown in the screenshot.

[vcf_sample.txt](https://github.com/user-attachments/files/18561378/vcf_sample.txt)

**Comment by cmdcolin** [comment]
Thanks for posting this @taprs 

I couldn't reproduce that error so maybe still gotta look a little more, if you have the full vcf can check that out



Random thing: I noticed that there are a lot of lines where no "ALT", or ALT is just a "."

I have not seen that much before. The VCF spec says that this means the ALT is "MISSING" (section 1.6.1) https://samtools.github.io/hts-specs/VCFv4.5.pdf

```
ALT — alternate base(s): Comma-separated list of alternate non-reference alleles. These alleles do not have to
be called in any of the samples. Each allele in this list must be one of: a non-empty String of bases (A,C,G,T,N;
case insensitive); the ‘*’ symbol (allele missing due to overlapping deletion); the MISSING value ‘.’ (no variant);
an angle-bracketed ID String (“<ID>”); the unspecified allele “<*>” as described in Section 5.5; or a breakend
replacement string as described in Section 5.4. If there are no alternative alleles, then the MISSING value must
be used. Tools processing VCF files are not required to preserve case in the allele String, except for IDs, which
are case sensitive. (String; no whitespace, commas, or angle-brackets are permitted in the ID String itself
```

**Comment by cmdcolin** [comment]
@taprs I did an update to the code and it has a number of improvements, potentially a fix for that crash also

**Comment by taprs** [comment]
Yes, the update fixed the VCF display issue I had, thanks a ton! It was weird because the error was gone if I sliced the vcf file...

As for the "missing" ALT values, we use them to have invariant sites present in the VCF and to distinguish them from sites with no data. This is useful for some popgen stuff.

**Comment by cmdcolin** [comment]
might be worth looking at UCSC version of 1000 genomes track

they cluster based on a 'central' variant in the view. I believe the reasoning is this can capture the 'local haplotype' similarity in a region

![Image](https://github.com/user-attachments/assets/9d28db52-635f-48d7-af66-34757a752b7f)

when haplotypes are actually plotted with a program like haploblocker you can see this quite tangibly that in the center of a screen, you can cluster haploblocks, but they get shuffled as you move away due

![Image](https://github.com/user-attachments/assets/bb234250-0ea7-49c6-8ddb-071f0d4d0923)



another interesting example: protein haplotypes from haplosaurus on ensemble

![Image](https://github.com/user-attachments/assets/fe7cb6f9-8a18-46db-855b-a7cce92403ee)

**Comment by cmdcolin** [comment]
we might want to 'improve' the display of minigraph-catus displays for example

example: screenshot showing all vs all PAF, minigraph cactus MAF, and minigraph cactus VCF. the "red" color was chosen for non-major alt alleles. probably a less severe color should be chosen...maybe

![Image](https://github.com/user-attachments/assets/2ffa6471-447c-4cbf-b9d1-de299037a2c3)

**Comment by cmdcolin** [comment]
probably gonna close this for now. i think we have nailed down most of the things we wanted  to get done. there are more directions we could go in...probably main ones would be 

- haploblocker
- more tutorials. some here https://jbrowse.org/jb2-staging/docs/tutorials/analyze_trio/
- multi-colored sidebar labels
- better pangenome representations (pangenome vcf look crazy generally with lots of red as seen above)


we'll incrementally work on those
